runtime/jcl/common/thread.cpp

```diff
@@ -587,33 +596,44 @@ Java_java_lang_VirtualThread_notifyJvmtiMountEnd(JNIEnv *env, jobject thread, jb
 void JNICALL
 Java_java_lang_VirtualThread_notifyJvmtiUnmountBegin(JNIEnv *env, jobject thread, jboolean lastUnmount)
 {
-}
+	J9VMThread *currentThread = (J9VMThread *)env;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
 
-/* private native void notifyJvmtiUnmountEnd(boolean lastUnmount); */
-void JNICALL
-Java_java_lang_VirtualThread_notifyJvmtiUnmountEnd(JNIEnv *env, jobject thread, jboolean lastUnmount)
-{
-	if (lastUnmount) {
-		J9VMThread *currentThread = (J9VMThread *)env;
-		J9JavaVM *vm = currentThread->javaVM;
-		J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	vmFuncs->internalEnterVMFromJNI(currentThread);
 
+	omrthread_monitor_enter(vm->liveVirtualThreadListMutex);
+	/* If this virtual thread is being inspected, do not allow it to die. */
+	j9object_t threadObj = J9_JNI_UNWRAP_REFERENCE(thread);
+	while (0 != J9OBJECT_U64_LOAD(currentThread, threadObj, vm->virtualThreadInspectorCountOffset)) {
+		VM_VMHelpers::pushObjectInSpecialFrame(currentThread, threadObj);
+		vmFuncs->internalExitVMToJNI(currentThread);
+		omrthread_monitor_wait(vm->liveVirtualThreadListMutex);
 		vmFuncs->internalEnterVMFromJNI(currentThread);
+		threadObj = VM_VMHelpers::popObjectInSpecialFrame(currentThread);
+	}
 
-		omrthread_monitor_enter(vm->liveVirtualThreadListMutex);
+	if (lastUnmount) {
 		if (NULL != vm->liveVirtualThreadList) {
-			j9object_t threadObj = J9_JNI_UNWRAP_REFERENCE(thread);
 			j9object_t threadPrev = J9OBJECT_OBJECT_LOAD(currentThread, threadObj, vm->virtualThreadLinkPreviousOffset);
 			j9object_t threadNext = J9OBJECT_OBJECT_LOAD(currentThread, threadObj, vm->virtualThreadLinkNextOffset);
 
 			/* Remove thread from list. The root will never be removed. */
 			J9OBJECT_OBJECT_STORE(currentThread, threadPrev, vm->virtualThreadLinkNextOffset, threadNext);
 			J9OBJECT_OBJECT_STORE(currentThread, threadNext, vm->virtualThreadLinkPreviousOffset, threadPrev);
 		}
-		omrthread_monitor_exit(vm->liveVirtualThreadListMutex);
-
-		vmFuncs->internalExitVMToJNI(currentThread);
 	}
+
+	omrthread_monitor_notify_all(vm->liveVirtualThreadListMutex);
```

Comment:
Is there a reason to add the `omrthread_monitor_notify_all()` call here ?

Reply:
More than one thread can be blocked on the `liveVirtualThreadListMutex`. This notify will make sure all blocked threads are incrementally woken up.

Reply:
They seems to be blocked on this mutex waiting for `virtualThreadInspectorCountOffset` to be 0. I don't see `virtualThreadInspectorCountOffset` changed in this function.

Reply:
It is incremented in `getVMThread` and decremented in `releaseVMThread`. I would suggest viewing the existing `J9VMThread->inspectorCount` changes. The new changes are a replica of them.

Reply:
I understand why releaseVMThread() called omrthread_monitor_notify_all(vm->liveVirtualThreadListMutex). It changes the inspector count and it is inside if (0 == vthreadInspectorCount) {

There was a typo in my previous comment. By virtualThreadInspectorCountOffset I mean virtualThreadInspectorCount

Reply:
After our discussion during lunch, I'm convinced that this notify is not needed.

Reply:
Updated.